### PR TITLE
choice observers, prep

### PR DESCRIPTION
**Pull request by nickchapman-da**
This PR contains preparatory work for _choice-observers_ #7709. The changes are simple, but wide ranging through the code base. They stop short of adding the new surface syntax in DAML. This will be done in a following PR.

These changes add support for _choice-observers_ to:

- Haskell AST -- optional `chcObservers` Expr
- Haskell `featureChoiceObservers`
- Haskell LF decode & encode
- Haskell type checker
- daml_lf_1.proto -- optional `observers` field in TemplateChoice message
- Scala AST  -- optional `choiceObservers` Expr
- Scala LF decode & encode
- Scala LF parser -- extend syntax for optional `observers` on `choice`s
- Scala LF type checker
- Scala speedy compiler -- pay attention to optional `choiceObservers`
- Scala speedy builtin `SBUBeginExercise` -- process choice-observers argument
- Scala engine -- construct transaction with choice-observers
- Scala transaction Node/NodeInfo
- Scala transaction version -- `minChoiceObservers`
- Scala transaction coder (encode & decode)
- transaction.proto -- new optional, repeated `observers` field

Things to be done in following PRs:

- Update specs: `daml-lf-1.rst` and `transaction.rst`
- Extend the surface syntax of DAML for choice-observers.
- Testing of which parties are informed. (Such tests will be easier using the surface syntax).

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/protobuf/com/daml/lf/transaction.proto
Comment (hurryabit) [inline daml-lf/transaction/src/main/protobuf/com/daml/lf/transaction.proto:97]: We should document that this need to be empty for whatever is the current version and below. We also need to agree if we want to call this `observers` or `informees`. cc @remyhaemmerle-da @bame-da @andreaslochbihler-da

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:232]: When computing the witnesses of an exercise node, you'll need to consider the template observers and the choice observers. Thus, I would prefer to call this field `choiceObservers` (or `informees` if we go with that) to avoid any confusion.

@remyhaemmerle-da Do we need to make this field optional for the sake of backward compatibility?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/NodeInfo.scala
Comment (hurryabit) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/NodeInfo.scala:77]: Why are we adding the observers for a non-consuming choice? That's the sort of confusion I was referring to above.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:363]:
```suggestion
          observers.foreach(checkExpr(_, TParties))
```
Don't need the comment anymore.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/traversable/TypeTraversable.scala:192]:
```suggestion
        observers.foreach(foreach(_, f))
```

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/ModParser.scala
Comment (remyhaemmerle-da) [inline daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/ModParser.scala:144]: If you do not plan to fix the parser in this PR, please add a FIXME with a github issue.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala
Comment (remyhaemmerle-da) [inline daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:191]: We probably have to add the `observers` to the argument of the definition. 
The question is if we should provide a default value. I think `Set.empty[String]` as a default value make sens.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:220]: We should check the version here ! 

You have to add in `TransactionVersions`: 
```scala
private[transaction] val minChoiceObserver = TransactionVersion("11")
```

And somewhere around here: 
```scala
if (transactionVersion precedes minChoiceObserver && ne.observers.nonEmpty) 
  throw  EncodeError(
                s"As of version $minNoControllers, the observer of an exercise node _must_ be empty",
  )
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:424]: We should check the version !

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala:662]: We shoudl check the version here as well.

**Review by remyhaemmerle-da — COMMENTED** [review]
Looks promising.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala
Comment (hurryabit) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala:130]: I'm in favor of preserving the optionality.

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala
Comment (hurryabit) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:191]: I'd prefer to not have a default value. I think they are just too dangerous.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs
Comment (hurryabit) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs:814]:
```suggestion
    -- ^ The observers of the choice. When they are present, they have type @List Party@.
```

File: compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs
Comment (hurryabit) [inline compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/DecodeV1.hs:322]:
```suggestion
    <*> traverse decodeExpr templateChoiceObservers
```
and drop the function below.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs
Comment (hurryabit) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs:663]: Rename `observers` above to `mbObservers` and
```suggestion
  introExprVar param paramType $ do
    checkExpr controllers (TList TParty)
    whenJust mbObservers $ \observers -> do
      -- check version is recent enough here
      checkExpr observers (TList TParty)
```
and drop `checkOptionalChoiceObservers` below.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs
Comment (hurryabit) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:658]: Let's do that in a separate PR.

File: daml-lf/language/src/test/scala/com/digitalasset/daml/lf/language/AstSpec.scala
Comment (hurryabit) [inline daml-lf/language/src/test/scala/com/digitalasset/daml/lf/language/AstSpec.scala:128]: We will probably need a new test for the `Some` case here as well.

File: daml-lf/parser/src/test/scala/com/digitalasset/daml/lf/testing/parser/ParsersSpec.scala
Comment (hurryabit) [inline daml-lf/parser/src/test/scala/com/digitalasset/daml/lf/testing/parser/ParsersSpec.scala:478]: We'll need a test for the `Some` case here as well.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala
Comment (hurryabit) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Node.scala:239]: I would not speak of "additional choice observers". It should either be "choice observers" or "additional informees".

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoDivulgenceSpec.scala
Comment (hurryabit) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/dao/JdbcLedgerDaoDivulgenceSpec.scala:59]: Given that this test is about divulgence and choice observers are too, we'll need some new tests here.

**Review by hurryabit — COMMENTED** [review]
I've left a mixture of wining about missing tests and Haskell pro tips. 😃

File: daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala
Comment (remyhaemmerle-da) [reply daml-lf/transaction-test-lib/src/main/scala/lf/transaction/test/TransactionBuilder.scala:191]: Note that this is a library for writing test, and it may used by external code. 
That said, I am OK with no default value. In any case the `observers` should be an argument of the method.

**Comment by nickchapman-da** [comment]
@Remy @Martin I've worked though all your comments (including the 2nd set from Martin). I tried not to miss anything. When I haven't made the change already, I added a NICK/FIXME marker in the code. I updated the desc to say what remains to be done before the first PR can be extracted.

One question:
- Should choice observers be added to the informees for a non-consuming choice?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:926]: Given that we need to change it anyway, let's spell it out. I also think speaking of "contract observers" and "choice observers" is somewhat inconsistent. It should either be "template observers" and "choice observers" or "contract observers" and "exercise observers". I heavily prefer the former.
```suggestion
      val templateObservers = extractParties(args.get(4))
```

**Comment by andreaslochbihler-da** [comment]
>     * Should choice observers be added to the informees for a non-consuming choice?

Yes. @oggy- will update the ledger model accordingly once all this has landed.

**Comment by remy** [comment]
Wrong "@remy". I'll unsubscribe myself…

**Comment by nickchapman-da** [comment]
> > ```
> > * Should choice observers be added to the informees for a non-consuming choice?
> > ```
> 
> Yes. @oggy- will update the ledger model accordingly once all this has landed.

Sorry. I didn't ask the question I intended!
Of course _choice observers_ must be _informees_ for a *non-consuming* choice. That's their purpose.
I meant to ask: should they be _informees_ of a *consuming* choice? I think the answer is yes.

**Comment by nickchapman-da** [comment]
> Wrong "@remy". I'll unsubscribe myself…

Sorry. I meant @remyhaemmerle-da

**Comment by andreaslochbihler-da** [comment]
> I meant to ask: should they be _informees_ of a _consuming_ choice? I think the answer is yes.

Yes. IIRC choice observers can be specified for each choice individually. So if the choice observers were not added to the informees of a consuming exercise, then specifying them on a consuming choice (which is determined statically in DAML) would be pointless.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs
Comment (hurryabit) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs:133]: Can you please leave a comment
```
- TODO Change as part of #7139
```
(like above) and add choice observers to that issue.

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala
Comment (nickchapman-da) [reply daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala:42]: For templates, "observers" is lexed as an identifier, not a keyword, and so requires additional `,`s to separate the various fields:
```
      (Id("precondition") ~> expr) ~
      (`,` ~> Id("signatories") ~> expr) ~
      (`,` ~> Id("observers") ~> expr) ~
      (`,` ~> Id("agreement") ~> expr) ~
```

whereas for choices we have this keyword style: `by`, `to`, which doesn't require commas.
(this is why I went for a new `ob` keyword)

```
      `by` ~ expr ~
      opt(`ob` ~> expr) ~
      `to` ~ expr ^^ {
```

If you want to move to the identifier style, then we should probably also change keyword `by` to identifier `controllers` to be consistent, and so require commas between all the fields.

Comment (nickchapman-da) [reply daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala]: I pushed a change to use identifier "observers" syntax (and leaving the `by` keyword as it is)
so you can see if you are happy with what it looks like
or decide how you want it to change
@hurryabit @remyhaemmerle-da

Comment (hurryabit) [reply daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala]: I'd be fine with moving from `by` to `controllers` and adding some commas. Could we do that in a prep PR?

Comment (nickchapman-da) [reply daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Lexer.scala]: done

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs
Comment (nickchapman-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs:134]: done

**Review by hurryabit — COMMENTED** [review]
I'm happy with the changes on the Haskell side and the API changes on the Scala side. @remyhaemmerle-da can you please approve this PR once you're happy with the rest.

File: daml-lf/spec/daml-lf-1.rst
Comment (remyhaemmerle-da) [inline daml-lf/spec/daml-lf-1.rst:4169]: :+1:

**Review by remyhaemmerle-da — APPROVED** [review]
Looks good to me.

**Review by remyhaemmerle-da — APPROVED** [review]
I am fine with merging the PR such as.

**Pull request merged**
